File: gherkin/dart/lib/src/language/Token.dart
import 'package:gherkin/core.dart';
import 'package:gherkin/extensions.dart';

import 'GherkinLineSpan.dart';
import 'IGherkinDialect.dart';
import 'IGherkinLine.dart';
import 'Location.dart';
import 'TokenType.dart';

class Token implements INullSafetyObject
{
  static Token get empty => _InvalidToken();

  IGherkinLine line = IGherkinLine.empty;

  Location location = Location.empty;

  TokenType matchedType = TokenType.None;

  String matchedKeyword = Strings.empty;

  String matchedText = Strings.empty;

  Iterable<GherkinLineSpan> matchedItems = const <GherkinLineSpan>[];

  int matchedIndent = 0;

  IGherkinDialect matchedGherkinDialect = IGherkinDialect.empty;

  bool get isEof => line.isEof;

  String get tokenValue => isEof ? 'EOF' : line.getLineText(-1);

  Token(this.line, this.location);

  void detach() {
    if (line.isNotEof) {
      line.detach();
    }
  }

  @override
  String toString() =>
      '${matchedType.name}: $matchedKeyword/$matchedText';

  @override
  bool get isEmpty => false;/*line.isEmpty
      && location.isEmpty
      && matchedKeyword.isEmpty
      && matchedText.isEmpty
      && matchedItems.isEmpty
      && matchedGherkinDialect.isEmpty;*/

  @override
  bool get isNotEmpty => !isEmpty;
}

/// Convenience implementation of an invalid [Token] instance.
class _InvalidToken extends Token
{
  _InvalidToken() : super(IGherkinLine.empty, Location.empty) {
    matchedGherkinDialect = IGherkinDialect.empty;
    matchedType = TokenType.None;
    matchedKeyword = Strings.empty;
    matchedText = Strings.empty;
    matchedItems = const <GherkinLineSpan>[];
    matchedIndent = 0;
  }

  @override
  bool get isEmpty => true;
}

File: gherkin/java/src/main/java/io/cucumber/gherkin/Token.java
package io.cucumber.gherkin;

import io.cucumber.messages.types.StepKeywordType;

import java.util.List;

class Token {
    public final IGherkinLine line;
    public Parser.TokenType matchedType;
    public String matchedKeyword;
    public String matchedText;
    public List<GherkinLineSpan> matchedItems;
    public int matchedIndent;
    public GherkinDialect matchedGherkinDialect;
    public StepKeywordType keywordType;
    public Location location;

    public Token(IGherkinLine line, Location location) {
        this.line = line;
        this.location = location;
    }

    public boolean isEOF() {
        return line == null;
    }

    public void detach() {
        if (line != null)
            line.detach();
    }

    public String getTokenValue() {
        return isEOF() ? "EOF" : line.getLineText(-1);
    }

    @Override
    public String toString() {
        return String.format("%s: %s/%s", matchedType, matchedKeyword, matchedText);
    }
}
